### ISSUE TITLE
Unknown type name ‘clockid_t’ error when building only Null platform

### ISSUE THREAD

**Issue by daljit46** [glfw/glfw#2402]
I've encountered a compilation issue when trying to build GLFW from source on Linux platforms (as part of a simple project that uses [dawn](https://dawn.googlesource.com/dawn)). The build instructions I have used are below. 
I've tested this on Ubuntu 22.04 LTS with GCC 11.4.0. 

```
git clone https://github.com/glfw/glfw
cd glfw
cmake -B build -G Ninja -DGLFW_USE_X11=OFF
cmake --build build
```
The project successfully configures, but there is a compilation error during the build stage. Someone on the dawn matrix channel pointed out that the following issue may be relevant https://github.com/floooh/sokol/issues/852.

Full compilation output:
```
[1/90] Building C object src/CMakeFiles/glfw.dir/platform.c.o
FAILED: src/CMakeFiles/glfw.dir/platform.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/platform.c.o -MF src/CMakeFiles/glfw.dir/platform.c.o.d -o src/CMakeFiles/glfw.dir/platform.c.o -c /home/ds23/Documents/Dev/glfw/src/platform.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/platform.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[2/90] Building C object src/CMakeFiles/glfw.dir/null_joystick.c.o
FAILED: src/CMakeFiles/glfw.dir/null_joystick.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/null_joystick.c.o -MF src/CMakeFiles/glfw.dir/null_joystick.c.o.d -o src/CMakeFiles/glfw.dir/null_joystick.c.o -c /home/ds23/Documents/Dev/glfw/src/null_joystick.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/null_joystick.c:29:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[3/90] Building C object src/CMakeFiles/glfw.dir/posix_module.c.o
FAILED: src/CMakeFiles/glfw.dir/posix_module.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -MD -MT src/CMakeFiles/glfw.dir/posix_module.c.o -MF src/CMakeFiles/glfw.dir/posix_module.c.o.d -o src/CMakeFiles/glfw.dir/posix_module.c.o -c /home/ds23/Documents/Dev/glfw/src/posix_module.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/posix_module.c:29:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[4/90] Building C object src/CMakeFiles/glfw.dir/posix_thread.c.o
FAILED: src/CMakeFiles/glfw.dir/posix_thread.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -MD -MT src/CMakeFiles/glfw.dir/posix_thread.c.o -MF src/CMakeFiles/glfw.dir/posix_thread.c.o.d -o src/CMakeFiles/glfw.dir/posix_thread.c.o -c /home/ds23/Documents/Dev/glfw/src/posix_thread.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/posix_thread.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[5/90] Building C object src/CMakeFiles/glfw.dir/vulkan.c.o
FAILED: src/CMakeFiles/glfw.dir/vulkan.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/vulkan.c.o -MF src/CMakeFiles/glfw.dir/vulkan.c.o.d -o src/CMakeFiles/glfw.dir/vulkan.c.o -c /home/ds23/Documents/Dev/glfw/src/vulkan.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/vulkan.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[6/90] Building C object src/CMakeFiles/glfw.dir/null_init.c.o
FAILED: src/CMakeFiles/glfw.dir/null_init.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/null_init.c.o -MF src/CMakeFiles/glfw.dir/null_init.c.o.d -o src/CMakeFiles/glfw.dir/null_init.c.o -c /home/ds23/Documents/Dev/glfw/src/null_init.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/null_init.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[7/90] Building C object src/CMakeFiles/glfw.dir/osmesa_context.c.o
FAILED: src/CMakeFiles/glfw.dir/osmesa_context.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/osmesa_context.c.o -MF src/CMakeFiles/glfw.dir/osmesa_context.c.o.d -o src/CMakeFiles/glfw.dir/osmesa_context.c.o -c /home/ds23/Documents/Dev/glfw/src/osmesa_context.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/osmesa_context.c:34:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[8/90] Building C object src/CMakeFiles/glfw.dir/egl_context.c.o
FAILED: src/CMakeFiles/glfw.dir/egl_context.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/egl_context.c.o -MF src/CMakeFiles/glfw.dir/egl_context.c.o.d -o src/CMakeFiles/glfw.dir/egl_context.c.o -c /home/ds23/Documents/Dev/glfw/src/egl_context.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/egl_context.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[9/90] Building C object src/CMakeFiles/glfw.dir/context.c.o
FAILED: src/CMakeFiles/glfw.dir/context.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/context.c.o -MF src/CMakeFiles/glfw.dir/context.c.o.d -o src/CMakeFiles/glfw.dir/context.c.o -c /home/ds23/Documents/Dev/glfw/src/context.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/context.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[10/90] Building C object src/CMakeFiles/glfw.dir/init.c.o
FAILED: src/CMakeFiles/glfw.dir/init.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/init.c.o -MF src/CMakeFiles/glfw.dir/init.c.o.d -o src/CMakeFiles/glfw.dir/init.c.o -c /home/ds23/Documents/Dev/glfw/src/init.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/init.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[11/90] Building C object src/CMakeFiles/glfw.dir/null_monitor.c.o
FAILED: src/CMakeFiles/glfw.dir/null_monitor.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/null_monitor.c.o -MF src/CMakeFiles/glfw.dir/null_monitor.c.o.d -o src/CMakeFiles/glfw.dir/null_monitor.c.o -c /home/ds23/Documents/Dev/glfw/src/null_monitor.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/null_monitor.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[12/90] Building C object src/CMakeFiles/glfw.dir/window.c.o
FAILED: src/CMakeFiles/glfw.dir/window.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/window.c.o -MF src/CMakeFiles/glfw.dir/window.c.o.d -o src/CMakeFiles/glfw.dir/window.c.o -c /home/ds23/Documents/Dev/glfw/src/window.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/window.c:31:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[13/90] Building C object src/CMakeFiles/glfw.dir/monitor.c.o
FAILED: src/CMakeFiles/glfw.dir/monitor.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/monitor.c.o -MF src/CMakeFiles/glfw.dir/monitor.c.o.d -o src/CMakeFiles/glfw.dir/monitor.c.o -c /home/ds23/Documents/Dev/glfw/src/monitor.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/monitor.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[14/90] Building C object src/CMakeFiles/glfw.dir/null_window.c.o
FAILED: src/CMakeFiles/glfw.dir/null_window.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/null_window.c.o -MF src/CMakeFiles/glfw.dir/null_window.c.o.d -o src/CMakeFiles/glfw.dir/null_window.c.o -c /home/ds23/Documents/Dev/glfw/src/null_window.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/null_window.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[15/90] Building C object src/CMakeFiles/glfw.dir/posix_time.c.o
FAILED: src/CMakeFiles/glfw.dir/posix_time.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -MD -MT src/CMakeFiles/glfw.dir/posix_time.c.o -MF src/CMakeFiles/glfw.dir/posix_time.c.o.d -o src/CMakeFiles/glfw.dir/posix_time.c.o -c /home/ds23/Documents/Dev/glfw/src/posix_time.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/posix_time.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
/home/ds23/Documents/Dev/glfw/src/posix_time.c: In function ‘_glfwPlatformInitTimer’:
/home/ds23/Documents/Dev/glfw/src/posix_time.c:44:31: error: ‘CLOCK_REALTIME’ undeclared (first use in this function)
   44 |     _glfw.timer.posix.clock = CLOCK_REALTIME;
      |                               ^~~~~~~~~~~~~~
/home/ds23/Documents/Dev/glfw/src/posix_time.c:44:31: note: each undeclared identifier is reported only once for each function it appears in
/home/ds23/Documents/Dev/glfw/src/posix_time.c:49:9: warning: implicit declaration of function ‘clock_gettime’ [-Wimplicit-function-declaration]
   49 |     if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0)
      |         ^~~~~~~~~~~~~
/home/ds23/Documents/Dev/glfw/src/posix_time.c:49:23: error: ‘CLOCK_MONOTONIC’ undeclared (first use in this function)
   49 |     if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0)
      |                       ^~~~~~~~~~~~~~~
[16/90] Building C object src/CMakeFiles/glfw.dir/input.c.o
FAILED: src/CMakeFiles/glfw.dir/input.c.o 
/usr/bin/cc  -I/home/ds23/Documents/Dev/glfw/include -I/home/ds23/Documents/Dev/glfw/src -I/home/ds23/Documents/Dev/glfw/build/src -fPIC -Wall -std=c99 -Wdeclaration-after-statement -MD -MT src/CMakeFiles/glfw.dir/input.c.o -MF src/CMakeFiles/glfw.dir/input.c.o.d -o src/CMakeFiles/glfw.dir/input.c.o -c /home/ds23/Documents/Dev/glfw/src/input.c
In file included from /home/ds23/Documents/Dev/glfw/src/platform.h:190,
                 from /home/ds23/Documents/Dev/glfw/src/internal.h:331,
                 from /home/ds23/Documents/Dev/glfw/src/input.c:30:
/home/ds23/Documents/Dev/glfw/src/posix_time.h:38:5: error: unknown type name ‘clockid_t’
   38 |     clockid_t   clock;
      |     ^~~~~~~~~
[22/90] Building C object examples/CMakeFiles/offscreen.dir/offscreen.c.o
ninja: build stopped: subcommand failed.

```

**Comment by GamesTrap** [comment]
After applying the following changes I am able to compile again using GCC 13.2.0

```diff
diff --git a/src/posix_time.c b/src/posix_time.c
index caed678e..a16552d5 100644
--- a/src/posix_time.c
+++ b/src/posix_time.c
@@ -27,6 +27,8 @@
 // It is fine to use C99 in this file because it will not be built with VS
 //========================================================================
 
+#define _POSIX_C_SOURCE 199309L
+
 #include "internal.h"
 
 #if defined(GLFW_BUILD_POSIX_TIMER)
diff --git a/src/posix_time.h b/src/posix_time.h
index 94374adb..899d058b 100644
--- a/src/posix_time.h
+++ b/src/posix_time.h
@@ -30,6 +30,7 @@
 #include <stdint.h>
 #include <time.h>
 
+typedef int clockid_t;
 
 // POSIX-specific global timer data
 //

```

**Comment by elmindreda** [comment]
This should be fixed now with d9709904fec956524d4e9032a7fa8d0550d5d49b. Thank you for reporting this!

@daljit46 Please let me know if you want your credit line changed to something else.
